File: rewrite/web/game/state.js
// A game is in one of these six states, only continuing from the abnormal states of 
// promotion, reverse promotion, and check.
export const State = {
    NORMAL: 0,
    PROMOTION: 1,
    CHECK: 2,
    CHECKMATE: 3,
    DRAW: 4,
    CONCEDED: 5,
    TIME_OVER: 6,
    REVERSE_PROMOTION: 7
}

function StateDef(name) {
    this.name = name
}

export const States = [
    new StateDef('Normal'),
    new StateDef('Promotion'),
    new StateDef('Check'),
    new StateDef('Checkmate'),
    new StateDef('Draw'),
    new StateDef('Conceded'),
    new StateDef('Time Over'),
    new StateDef('Reverse Promotion')
]

File: rules/state.go
package rules

import "strconv"

// State represents the current state of the game, such as check, promotion needed, or draw.
type State int

const (
	Normal State = iota
	Promotion
	Check
	Checkmate
	Draw
	Conceded // this and the following aren't used by the rules package
	TimeOver
	ReversePromotion
	NoState
)

func (a State) String() string {
	switch a {
	case Normal:
		return "normal"
	case Promotion:
		return "promotion"
	case Check:
		return "check"
	case Checkmate:
		return "checkmate"
	case Draw:
		return "draw"
	case Conceded:
		return "conceded"
	case TimeOver:
		return "time over"
	case ReversePromotion:
		return "reverse promotion"
	}
	return strconv.Itoa(int(a))
}
